File: python/102.py
# 二叉树的层序遍历
# Binary Tree Level Order Traversal

from common import List, Optional, TreeNode, create_binary_tree


class Solution:
    def levelOrder(self, root: Optional[TreeNode]) -> List[List[int]]:
        result: List[List[int]] = []

        if not root:
            return result

        from collections import deque

        queue = deque([(root, 0)])

        while len(queue):
            node, level = queue.popleft()
            if node is None:
                continue
            if len(result) < level + 1:
                result.append([])
            result[level].append(node.val)
            queue.append((node.left, level + 1))
            queue.append((node.right, level + 1))

        return result


if __name__ == "__main__":
    test_cases = [
        [3, 9, 20, None, None, 15, 7],
        [1],
        [],
    ]
    sol = Solution()
    for case in test_cases:
        root = create_binary_tree(case)
        print(sol.levelOrder(root))

File: cpp/102.cpp
// 二叉树的层序遍历
// Binary Tree Level Order Traversal

#include "common.h"

class Solution
{
public:
    std::vector<std::vector<int>> levelOrder(TreeNode *root)
    {
        std::vector<std::vector<int>> result = {};
        if (root == nullptr)
        {
            return result;
        }

        std::deque<std::pair<TreeNode *, int>> queue;
        queue.push_back({root, 0});

        while (!queue.empty())
        {
            auto [node, level] = queue.front();
            queue.pop_front();
            if (node == nullptr)
            {
                continue;
            }
            if (level >= result.size())
            {
                result.resize(level + 1);
            }
            result[level].push_back(node->val);
            queue.push_back({node->left, level + 1});
            queue.push_back({node->right, level + 1});
        }

        return result;
    }
};

int main()
{
    std::vector<std::vector<int>> test_cases = {
        {3, 9, 20, NULL_VALUE, NULL_VALUE, 15, 7},
        {1},
        {},
    };

    Solution sol;
    for (const auto &test_case : test_cases)
    {
        auto root = create_binary_tree(test_case);
        auto result = sol.levelOrder(root);
        for (const auto &level : result)
        {
            print_vector(level);
        }
    }
}
